arangosh/Backup/BackupFeature.cpp

```diff
@@ -0,0 +1,509 @@
+#include <regex>
+#include <thread>
+
+#include <velocypack/Iterator.h>
+
+#include "Basics/StaticStrings.h"
+#include "Basics/StringUtils.h"
+#include "Basics/VelocyPackHelper.h"
+#include "Logger/Logger.h"
+#include "ProgramOptions/ProgramOptions.h"
+#include "SimpleHttpClient/SimpleHttpClient.h"
+#include "SimpleHttpClient/SimpleHttpResult.h"
+
+namespace {
+
+/// @brief name of the feature to report to application server
+constexpr auto FeatureName = "Backup";
+
+constexpr auto OperationCreate = "create";
+constexpr auto OperationDelete = "delete";
+constexpr auto OperationList = "list";
+constexpr auto OperationRestore = "restore";
+std::unordered_set<std::string> const Operations = {OperationCreate, OperationDelete,
+                                                    OperationList, OperationRestore};
+
+/// @brief generic error for if server returns bad/unexpected json
+const arangodb::Result ErrorMalformedJsonResponse = {
+    TRI_ERROR_INTERNAL, "got malformed JSON response from server"};
+
+/// @brief check whether HTTP response is valid, complete, and not an error
+arangodb::Result checkHttpResponse(arangodb::httpclient::SimpleHttpClient& client,
+                                   std::unique_ptr<arangodb::httpclient::SimpleHttpResult> const& response) {
+  using arangodb::basics::StringUtils::itoa;
+  if (response == nullptr || !response->isComplete()) {
+    return {TRI_ERROR_INTERNAL,
+            "got invalid response from server: " + client.getErrorMessage()};
+  }
+  if (response->wasHttpError()) {
+    int errorNum = TRI_ERROR_INTERNAL;
+    std::string errorMsg = response->getHttpReturnMessage();
+    std::shared_ptr<arangodb::velocypack::Builder> bodyBuilder(response->getBodyVelocyPack());
+    arangodb::velocypack::Slice error = bodyBuilder->slice();
+    if (!error.isNone() && error.hasKey(arangodb::StaticStrings::ErrorMessage)) {
+      errorNum = error.get(arangodb::StaticStrings::ErrorNum).getNumericValue<int>();
+      errorMsg = error.get(arangodb::StaticStrings::ErrorMessage).copyString();
+    }
+    return {errorNum, "got invalid response from server: HTTP " +
+                          itoa(response->getHttpReturnCode()) + ": " + errorMsg};
+  }
+  return {TRI_ERROR_NO_ERROR};
+}
+
+arangodb::Result getUptime(arangodb::httpclient::SimpleHttpClient& client, double& uptime) {
+  arangodb::Result result;
+
+  std::string const url = "/_admin/statistics";
+  std::unique_ptr<arangodb::httpclient::SimpleHttpResult> response(
+      client.request(arangodb::rest::RequestType::GET, url, nullptr, 0));
+  auto check = ::checkHttpResponse(client, response);
+  if (check.fail()) {
+    return check;
+  }
+
+  // extract vpack body from response
+  std::shared_ptr<VPackBuilder> parsedBody;
+  try {
+    parsedBody = response->getBodyVelocyPack();
+  } catch (...) {
+    result.reset(::ErrorMalformedJsonResponse);
+    return result;
+  }
+  VPackSlice const resBody = parsedBody->slice();
+  TRI_ASSERT(resBody.isObject());
```

Comment:
I don't think that asserts are the right things to do with a remote provided input data - they must not be trusted.